alacritty_terminal/src/selection.rs

```diff
@@ -83,70 +92,119 @@ impl<L> SelectionRange<L> {
+        }
     }
 
-    pub fn lines(point: Point<usize>) -> Selection {
-        Selection::Lines { region: Range { start: point, end: point } }
+    pub fn lines(location: Point<usize>) -> Selection {
+        Self {
+            region: Range {
+                start: Anchor::new(location, Side::Left),
+                end: Anchor::new(location, Side::Right),
+            },
+            ty: SelectionType::Lines,
+        }
     }
 
     pub fn update(&mut self, location: Point<usize>, side: Side) {
-        let (_, end) = self.points_mut();
-        *end = location;
+        self.region.end.point = location;
+        self.region.end.side = side;
+    }
+
+    pub fn rotate(
+        mut self,
+        num_lines: usize,
+        num_cols: usize,
+        scrolling_region: &Range<Line>,
+        offset: isize,
+    ) -> Option<Selection> {
+        // Only rotate selection inside of scrolling region
+        let region_start = num_lines - scrolling_region.start.0;
+        let region_end = num_lines - scrolling_region.end.0;
+
+        let (mut start, mut end) = (&mut self.region.start, &mut self.region.end);
+        if Self::points_need_swap(start.point, end.point) {
+            mem::swap(&mut start, &mut end);
+        }
+
+        // Rotate start of selection
+        if (start.point.line < region_start || region_start == num_lines)
+            && start.point.line >= region_end
+        {
+            start.point.line = usize::try_from(start.point.line as isize + offset).unwrap_or(0);
 
-        if let Some((_, end_side)) = self.sides_mut() {
-            *end_side = side;
+            // If end is within the same region, delete selection once start rotates out
+            if start.point.line < region_end && end.point.line >= region_end {
+                return None;
+            }
+
+            // Clamp selection to start of region
+            if start.point.line >= region_start && region_start != num_lines {
+                if self.ty != SelectionType::Block {
+                    start.point.col = Column(0);
+                    start.side = Side::Left;
+                }
+                start.point.line = region_start - 1;
+            }
         }
-    }
 
-    pub fn rotate(&mut self, offset: isize) {
-        let (start, end) = self.points_mut();
-        start.line = usize::try_from(start.line as isize + offset).unwrap_or(0);
-        end.line = usize::try_from(end.line as isize + offset).unwrap_or(0);
+        // Rotate end of selection
+        if (end.point.line < region_start || region_start == num_lines)
+            && end.point.line >= region_end
+        {
+            end.point.line = usize::try_from(end.point.line as isize + offset).unwrap_or(0);
+
+            // Delete selection if end has overtaken the start
+            if end.point.line > start.point.line {
+                return None;
+            }
+
+            // Clamp selection to end of region
+            if end.point.line < region_end {
```

Comment:
How could you even trigger this thing?

Reply:
You can scroll in both directions in scrolling regions. Think about vim for example.